### PR TITLE
SUBMARINE-277. Support Spark Interpreter

### PR DESCRIPTION
**Pull request by luzhonghao**
### What is this PR for?
support spark interpreter for submarine

### What type of PR is it?
Feature

### What is the Jira issue?
https://issues.apache.org/jira/browse/SUBMARINE-277

### How should this be tested?
CI : https://travis-ci.org/luzhonghao/hadoop-submarine/builds/608108764

### Questions:
* Does the licenses files need update?No
* Is there breaking changes for older versions? No
* Does this needs documentation? No

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -119,6 +119,9 @@
     <zeppelin.version>0.9.0-SNAPSHOT</zeppelin.version>
     <jgit.version>5.5.1.201910021850-r</jgit.version>
     <atomix.version>3.0.0-rc4</atomix.version>
+    <libthrift.version>0.12.0</libthrift.version>
```
Comment (liuxunorg) [inline pom.xml]: Why need add `libthrift.version` ?

File: submarine-dist/src/assembly/distribution.xml [hunk 2]
```diff
@@ -135,6 +135,20 @@
         <include>python-interpreter-${project.version}-shade.jar</include>
       </includes>
     </fileSet>
+    <fileSet>
+      <directory>../submarine-workbench/interpreter/spark/scala-2.11</directory>
```
Comment (liuxunorg) [inline submarine-dist/src/assembly/distribution.xml]: Only support scala-2.11?

File: submarine-workbench/interpreter/spark-interpreter/pom.xml [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+
+            <plugin>
+                <artifactId>maven-enforcer-plugin</artifactId>
+                <executions>
+                    <execution>
+                        <id>enforce</id>
+                        <phase>none</phase>
+                    </execution>
+                </executions>
+            </plugin>
```
Comment (liuxunorg) [inline submarine-workbench/interpreter/spark-interpreter/pom.xml]: Need change to
```
      <plugin>
        <artifactId>maven-enforcer-plugin</artifactId>
        <configuration>
          <skip>false</skip>
        </configuration>
      </plugin>
```

File: submarine-workbench/interpreter/spark-interpreter/src/main/java/org/apache/submarine/interpreter/SparkInterpreter.java [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+import java.io.File;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Properties;
+
+public class SparkInterpreter  extends  InterpreterProcess{
```
Comment (liuxunorg) [inline submarine-workbench/interpreter/spark-interpreter/src/main/java/org/apache/submarine/interpreter/SparkInterpreter.java]: 1. Have two space.
2. `InterpreterProcess{` need add a spaces

File: submarine-workbench/interpreter/spark-interpreter/src/main/java/org/apache/submarine/interpreter/SparkInterpreter.java [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+    } catch (InterpreterException e) {
+      LOG.error(e.getMessage(), e);
+    }
+
+    return process;
+  }
+
+  @Override
+  public boolean test() {
+    return true;
```
Comment (liuxunorg) [inline submarine-workbench/interpreter/spark-interpreter/src/main/java/org/apache/submarine/interpreter/SparkInterpreter.java]: Need add real test code.

**Comment by luzhonghao** [comment]
> Please add Test Case.

Thanks for reviewing , done.

File: submarine-dist/src/assembly/distribution.xml (hunk 2 above)
Comment (luzhonghao) [reply submarine-dist/src/assembly/distribution.xml]: Currently , we only support scala-2.11

File: submarine-workbench/interpreter/spark-interpreter/src/main/java/org/apache/submarine/interpreter/SparkInterpreter.java (hunk 5 above)
Comment (luzhonghao) [reply submarine-workbench/interpreter/spark-interpreter/src/main/java/org/apache/submarine/interpreter/SparkInterpreter.java]: > Need add real test code.

Many thanks, I will fix it.

**Comment by liuxunorg** [comment]
@luzhonghao Need change PR title to 
SUBMARINE-277. [WIP]Support Spark Interpreter

PR title must is SUBMARINE-XXX. title

File: submarine-workbench/interpreter/spark-interpreter/pom.xml (hunk 3 above)
Comment (luzhonghao) [reply submarine-workbench/interpreter/spark-interpreter/pom.xml]: It seems that zeppelin‘s dependencies can not pass when the maven-enforcer-plugin is strictly set to false, I will try to fix it as possible.

File: pom.xml (hunk 1 above)
Comment (luzhonghao) [reply pom.xml]: Thanks to point it , the libthrift dependency is not needed for this PR, I will remove it.

File: submarine-workbench/interpreter/spark-interpreter/src/main/java/org/apache/submarine/interpreter/SparkInterpreter.java (hunk 4 above)
Comment (luzhonghao) [reply submarine-workbench/interpreter/spark-interpreter/src/main/java/org/apache/submarine/interpreter/SparkInterpreter.java]: Done

**Comment by luzhonghao** [comment]
> @luzhonghao Need change PR title to
> SUBMARINE-277. [WIP]Support Spark Interpreter
> 
> PR title must is SUBMARINE-XXX. title

Done

**Comment by luzhonghao** [comment]
@hhhizzz pls help me review , thanks

File: submarine-workbench/interpreter/spark-interpreter/src/main/java/org/apache/submarine/interpreter/SparkInterpreter.java [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+      return false;
+    }
+    boolean success = (result.message().get(0).getData().contains(
+        "+---+----+\n" +
+            "| _1|  _2|\n" +
+            "+---+----+\n" +
+            "|  1|   a|\n" +
+            "|  2|null|\n" +
+            "+---+----+"));
+    return success;
```
Comment (liuxunorg) [inline submarine-workbench/interpreter/spark-interpreter/src/main/java/org/apache/submarine/interpreter/SparkInterpreter.java]: maybe need add call `close()` ?

Comment (luzhonghao) [reply submarine-workbench/interpreter/spark-interpreter/src/main/java/org/apache/submarine/interpreter/SparkInterpreter.java]: nice catch , I will fix it.

Comment (luzhonghao) [reply submarine-workbench/interpreter/spark-interpreter/src/main/java/org/apache/submarine/interpreter/SparkInterpreter.java]: nice catch , I will fix it.

**Comment by liuxunorg** [comment]
@luzhonghao Thank you contribute this feature.
Will merge if no more comment.

**Pull request closed without merging**